cmd/egress-cni-plugin/cni/cni4_test.go

```diff
@@ -0,0 +1,201 @@
+	var actualIptablesRules []string
+
+	expectRouteDel := []string{"route del: {Ifindex: 2 Dst: 169.254.172.0/22 Src: <nil> Gw: <nil> Flags: [] Table: 0}"}
+	var actualRouteDel []string
+
+	expectRouteAdd := []string{
+		"route add: {Ifindex: 2 Dst: 169.254.172.1/32 Src: 169.254.172.100 Gw: <nil> Flags: [] Table: 0}",
+		"route add: {Ifindex: 2 Dst: 169.254.172.0/22 Src: 169.254.172.100 Gw: 169.254.172.1 Flags: [] Table: 0}",
+		"route add: {Ifindex: 100 Dst: 169.254.172.100/32 Src: <nil> Gw: <nil> Flags: [] Table: 0}"}
+	var actualRouteAdd []string
+
+	// setup the mock EXPECT
+	err := SetupAddExpectV4(context, &actualIptablesRules, &actualRouteAdd, &actualRouteDel)
+	assert.Nil(t, err)
+
+	// run egress plugin ADD action
+	err = CmdAddEgressV4(&context)
+	assert.Nil(t, err)
+
+	// confirm all iptables chain/rule are created
+	assert.EqualValuesf(t, expectIptablesRules, actualIptablesRules, "expected iptables chain/rule are not created")
+
+	// confirm automatically added route is removed
+	assert.EqualValuesf(t, expectRouteDel, actualRouteDel, "route are expected to removed")
+
+	// confirm routes are added in container and node
+	assert.EqualValuesf(t, expectRouteAdd, actualRouteAdd, "route are expected to added")
+
+	// the unit test write some output string not ends with '\n' and this cause go runner unable to interpret that a test was run.
+	// Adding a newline, keeps a clean output
+	fmt.Println()
+
+}
+
+func TestCmdDelEgressV4(t *testing.T) {
+	context := setupDelContextV4(gomock.NewController(t))
+
+	expectLinkDel := []string{"link del - name: v4if0"}
+	actualLinkDel := []string{}
+
+	expectIptablesDel := []string{
+		"nat POSTROUTING -s 169.254.172.100 -j CNI-E4 -m comment --comment unit-test-comment",
+		"clear chain nat CNI-E4",
+		"del chain nat CNI-E4"}
+	actualIptablesDel := []string{}
+
+	// run egress plugin DEL action
+	err := SetupDelExpectV4(context, &actualLinkDel, &actualIptablesDel)
+	assert.Nil(t, err)
+
+	err = CmdDelEgressV4(&context)
+	assert.Nil(t, err)
+
+	// confirm added interface is removed
+	assert.EqualValuesf(t, expectLinkDel, actualLinkDel, "interface are expected to deleted")
+
+	// confirm iptables chain/rules are deleted
+	assert.EqualValuesf(t, expectIptablesDel, actualIptablesDel, "iptables chain/rules are expected to deleted")
+}
+
+func setupAddContextV4(ctrl *gomock.Controller) Context {
+	ipt := mock_iptables.NewMockIptablesIface(ctrl)
+	return Context{
+		Procsys: mock_procsyswrapper.NewMockProcSys(ctrl),
+		Ns:      mock_nswrapper.NewMockNS(ctrl),
+		NsPath:  "/var/run/netns/cni-xxxx",
+		Ipam:    mock_ipamwrapper.NewMockIpam(ctrl),
+		Link:    mock_netlinkwrapper.NewMockNetLink(ctrl),
+		Veth:    mock_veth.NewMockVeth(ctrl),
+		NetConf: &NetConf{
+			NetConf: types.NetConf{
+				CNIVersion: "0.4.0",
+			},
+			NodeIP:         net.ParseIP("192.168.0.23"),
+			IfName:         EgressIPv4InterfaceName,
+			MTU:            "9001",
+			PluginLogFile:  "plugin.log",
```

Comment:
It can be changed to egress-plugin.log, but we could not use the global const here since global one usually starts with /var and local dev machine may not have directory.